File: src/main/java/com/learn/scaryCoders/dataStrucutres/DoublyLinkedList.java
package com.learn.scaryCoders.dataStrucutres;

import java.util.Iterator;
import java.util.Objects;

public class DoublyLinkedList<T> {
    private Node<T> head, tail;
    private int length;

    public DoublyLinkedList() {
    }

    public boolean add(T data) {
        if (Objects.isNull(data))
            return false;

        Node<T> node = new Node<>(data, null, null);
        if (head == null) {
            this.head = this.tail = node;
        } else {
            tail.next = node;
            node.prev = tail;
            tail = node;
        }
        length++;
        return true;
    }

    public boolean addAt(T data, int index) {
        if (Objects.isNull(data))
            return false;
        if (index > length && index < 0) {
            throw new IllegalArgumentException("index is out of bound");
        }

        Node tNode = new Node(data, null, null);
        if (head == tail || index == length - 1) {
            return add(data);
        }

        Node<T> temp = head;
        int count = 0;

        while (temp != null) {

            if (count == index) {
                if (temp.prev != null) {
                    tNode.prev = temp.prev;
                    tNode.next = temp;
                    temp.prev.next = tNode;
                    temp.prev = tNode;
                } else {
                    tNode.next = temp;
                    temp.prev = tNode;
                    head = tNode;
                }
                length++;
                return true;
            }
            count++;
            temp = temp.next;
        }
        return false;
    }


    public T remove(T data) {
        Node<T> temp = head;

        while (temp != null) {

            if (temp.data.equals(data)) {

                temp.prev.next = temp.next;
                temp.next.prev = temp.prev;
                length--;
                temp.prev = null;
                temp.next = null;
                return temp.data;
            }
            temp = temp.next;
        }
        return null;
    }

    public T removeAt(int index) {
        if (index > length - 1 && index < 0) {
            throw new IllegalArgumentException("Index is out of bounds");
        }

        Node<T> temp = head;
        int count = 0;

        while (temp != null) {
            if (count == index) {

                if (temp.prev != null && temp.next != null) {
                    temp.prev.next = temp.next;
                    temp.next.prev = temp.prev;
                } else {
                    if (temp.prev == null) {
                        head = temp.next;
                        head.prev = null;
                    } else {
                        tail = temp.prev;
                        tail.next = null;
                    }
                }
                temp.next = null;
                temp.prev = null;
                length--;
                return temp.data;
            }
            count++;
            temp = temp.next;
        }
        return null;
    }

    public T elementAtTail(){
        if(tail==null)
            return null;
        return tail.data;
    }

    public T elementAtHead(){
        if(head==null)
            return null;
        return head.data;
    }

    public boolean isEmpty() {
        return length == 0;
    }

    public int size() {
        return length;
    }

    public String print(){
        StringBuilder sb=new StringBuilder();
        Iterator<T> iterator = this.iterator();
        while(iterator.hasNext()){
            T data= iterator.next();
            sb.append(data+" ,");
        }
        return sb.toString();
    }

    public Iterator<T> iterator() {
        return new Iterator<T>() {
            Node<T> temp = head;
            private int len = length;

            @Override
            public boolean hasNext() {
                return len > 0;
            }

            @Override
            public T next() {
                T data = temp.data;
                temp = temp.next;
                len--;
                return data;
            }
        };
    }

    private class Node<T> {
        final T data;
        Node<T> next;
        Node<T> prev;

        Node(T data, Node<T> prev, Node<T> next) {
            this.data = data;
            this.prev = prev;
            this.next = next;
        }
    }

    public static void main(String s[]){
        DoublyLinkedList<String> dll = new DoublyLinkedList<>();
        dll.add("Ruksad");
        dll.add("Siddiqui");
        System.out.println(dll.print());
        dll.removeAt(1);
        System.out.println(dll.print());
        dll.addAt("Samuel",1);
        dll.addAt("Derric",0);

        System.out.println(dll.print());
    }
}

File: python/doubly_linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None


class Doubly_linked_list:
    def __init__(self):
        self.size = 0
        self.head = None
        self.tail = None


def print_doubly_linked_list(linked_list):
    if is_none(linked_list) or is_empty(linked_list):
        print("Please pass a non empty linked list")
        return

    temp = linked_list.head

    while temp.next is not None:
        print(temp.data, end=",")
        temp = temp.next
    print(temp.data)


def add_note_at(linked_list, data, location):
    node = Node(str(data))
    if is_empty(linked_list):
        linked_list.head = node
        linked_list.tail = node
        linked_list.size += 1
    else:
        temp = linked_list.head
        if location <= 0:
            location = 1
        elif location > linked_list.size:
            location = linked_list.size

        for i in range(1, location + 1):

            if i == location:
                if location == 1:
                    node.next = linked_list.head
                    linked_list.head.prev = node
                    linked_list.head = node
                elif location == linked_list.size:
                    node.prev = linked_list.tail
                    linked_list.tail.next = node
                    linked_list.tail = node
                else:
                    node.next = temp
                    node.prev = temp.prev
                    temp.prev.next = node
                    temp.prev = node
                linked_list.size += 1
            temp = temp.next


def is_none(linked_list):
    return linked_list == None


def is_empty(linked_list):
    return linked_list.size == 0


if __name__ == "__main__":
    dll = Doubly_linked_list()
    add_note_at(dll, "Mohammad", -1)
    add_note_at(dll, "Ruksad", -1)
    add_note_at(dll, "derric", 100)
    add_note_at(dll, "Siddiqui", -10)
    add_note_at(dll, "samuel", 2)
    print_doubly_linked_list(dll)
    add_note_at(dll, "samuel1", 2)
    print_doubly_linked_list(dll)
    add_note_at(dll, "samuel2", 3)
    print_doubly_linked_list(dll)
